crates/c-api/src/async.rs

```diff
@@ -0,0 +1,296 @@
+        let this = self.get_mut();
+        let cb = this.callback;
+        if cb(this.env) {
+            Poll::Ready(())
+        } else {
+            Poll::Pending
+        }
+    }
+}
+
+pub type wasmtime_func_async_continuation_callback_t = extern "C" fn(*mut c_void) -> bool;
+
+struct CallbackData {
+    env: *mut c_void,
+}
+unsafe impl Send for CallbackData {}
+
+async fn invoke_c_async_callback<'a>(
+    cb: wasmtime_func_async_callback_t,
+    data: CallbackData,
+    mut caller: Caller<'a, crate::StoreData>,
+    params: &'a [Val],
+    results: &'a mut [Val],
+) -> Result<()> {
+    // Convert `params/results` to `wasmtime_val_t`. Use the previous
+    // storage in `hostcall_val_storage` to help avoid allocations all the
+    // time.
+    let mut hostcall_val_storage = mem::take(&mut caller.data_mut().hostcall_val_storage);
+    debug_assert!(hostcall_val_storage.is_empty());
+    hostcall_val_storage.reserve(params.len() + results.len());
+    hostcall_val_storage.extend(params.iter().cloned().map(|p| wasmtime_val_t::from_val(p)));
+    hostcall_val_storage.extend((0..results.len()).map(|_| wasmtime_val_t {
+        kind: WASMTIME_I32,
+        of: wasmtime_val_union { i32: 0 },
+    }));
+    let (params, out_results) = hostcall_val_storage.split_at_mut(params.len());
+
+    // Invoke the C function pointer.
+    // The result will be a continutation which we will wrap in a Future.
+    let mut caller = wasmtime_caller_t { caller };
+    let mut trap = None;
+    let continuation = cb(
+        data.env,
+        &mut caller,
+        params.as_ptr(),
+        params.len(),
+        out_results.as_mut_ptr(),
+        out_results.len(),
+        &mut trap,
+    );
+    continuation.await;
+
+    if let Some(trap) = trap {
+        return Err(trap.error);
+    }
+
+    // Translate the `wasmtime_val_t` results into the `results` space
+    for (i, result) in out_results.iter().enumerate() {
+        unsafe {
+            results[i] = result.to_val();
+        }
+    }
+    // Move our `vals` storage back into the store now that we no longer
+    // need it. This'll get picked up by the next hostcall and reuse our
+    // same storage.
+    let mut v = mem::take(&mut hostcall_val_storage);
+    v.truncate(0);
+    caller.caller.data_mut().hostcall_val_storage = v;
+    Ok(())
+}
+
+unsafe fn c_async_callback_to_rust_fn(
+    callback: wasmtime_func_async_callback_t,
+    data: *mut c_void,
+    finalizer: Option<extern "C" fn(*mut std::ffi::c_void)>,
+) -> impl for<'a> Fn(
+    Caller<'a, crate::StoreData>,
+    &'a [Val],
+    &'a mut [Val],
+) -> Box<dyn Future<Output = Result<()>> + Send + 'a>
+       + Send
+       + Sync
+       + 'static {
+    let foreign = crate::ForeignData { data, finalizer };
+    move |caller, params, results| {
+        let _ = &foreign; // move entire foreign into this closure
+        let data = CallbackData { env: foreign.data };
```

Comment:
If you'd like, to avoid the `CallbackData` structure here, which I'm assuming is only required to satisfy `Send`, the `invoke_c_async_callback` function could take `&ForeignData` which I think is already `Send`

Reply:
I can't get that to work, which maybe means I'm doing something unsafe here...

The compiled tells me that the reference to the captured variables can't escape the closure. I'm not sure if Wasmtime ensures that the futures are completed before deleting the host function closures, from my debugging it seems the order is correct. Additionally c_void is not `Send` so I need the wrapper type. Other suggestions or am I doing something dangerous here?

Reply:
Ah no I see why rustc would say that, and makes sense. Sorry for the wild goose chase, this is good as-is!